user_tools/custom_speedup_factors/generate_speedup_factors.py

```diff
@@ -148,20 +150,41 @@
+if 'Scan orc ' in cpu_stage_totals and 'GpuScan orc ' in gpu_stage_totals:
+    scores_dict["BatchScanExec"] = str(round(cpu_stage_totals['Scan orc '] / gpu_stage_totals['GpuScan orc '], 2))
+    scores_dict["FileSourceScanExec"] = str(round(cpu_stage_totals['Scan orc '] / gpu_stage_totals['GpuScan orc '], 2))
+
+# Other operators
+if 'Project' in cpu_stage_totals and 'GpuProject' in gpu_stage_totals:
+    scores_dict["ProjectExec"] = str(round(cpu_stage_totals['Project'] / gpu_stage_totals['GpuProject'], 2))
+if 'Expand' in cpu_stage_totals and 'GpuExpand' in gpu_stage_totals:
+    scores_dict["ExpandExec"] = str(round(cpu_stage_totals['Expand'] / gpu_stage_totals['GpuExpand'], 2))
+if 'CartesianProduct' in cpu_stage_totals and 'GpuCartesianProduct' in gpu_stage_totals:
+    scores_dict["CartesianProductExec"] = str(round(cpu_stage_totals['CartesianProduct'] / gpu_stage_totals['GpuCartesianProduct'], 2))
 if 'Filter' in cpu_stage_totals and 'GpuFilter' in gpu_stage_totals:
     scores_dict["FilterExec"] = str(round(cpu_stage_totals['Filter'] / gpu_stage_totals['GpuFilter'], 2))
 if 'SortMergeJoin' in cpu_stage_totals and 'GpuShuffledHashJoin' in gpu_stage_totals:
-    scores_dict["SortExec"] = str(round(cpu_stage_totals['SortMergeJoin'] / gpu_stage_totals['GpuShuffledHashJoin'], 2))
+    scores_dict["SortMergeJoinExec"] = str(round(cpu_stage_totals['SortMergeJoin'] / gpu_stage_totals['GpuShuffledHashJoin'], 2))
 if 'BroadcastHashJoin' in cpu_stage_totals and 'GpuBroadcastHashJoin' in gpu_stage_totals:
     scores_dict["BroadcastHashJoinExec"] = str(round(cpu_stage_totals['BroadcastHashJoin'] / gpu_stage_totals['GpuBroadcastHashJoin'], 2))
 if 'Exchange' in cpu_stage_totals and 'GpuColumnarExchange' in gpu_stage_totals:
     scores_dict["ShuffleExchangeExec"] = str(round(cpu_stage_totals['Exchange'] / gpu_stage_totals['GpuColumnarExchange'], 2))
 if 'HashAggregate' in cpu_stage_totals and 'GpuHashAggregate' in gpu_stage_totals:
     scores_dict["HashAggregateExec"] = str(round(cpu_stage_totals['HashAggregate'] / gpu_stage_totals['GpuHashAggregate'], 2))
-if all(cpu_keys in cpu_stage_totals for cpu_keys in ('SortMergeJoin', 'Sort' )) and all(gpu_keys in gpu_stage_totals for gpu_keys in ('GpuShuffledHashJoin', 'GpuSort')):
-    scores_dict["SortMergeJoinExec"] = str(round((cpu_stage_totals['SortMergeJoin'] + cpu_stage_totals['Sort']) / (gpu_stage_totals['GpuShuffledHashJoin'] + gpu_stage_totals['GpuSort']), 2))
+    scores_dict["ObjectHashAggregateExec"] = str(round(cpu_stage_totals['HashAggregate'] / gpu_stage_totals['GpuHashAggregate'], 2))
+    scores_dict["SortAggregateExec"] = str(round(cpu_stage_totals['HashAggregate'] / gpu_stage_totals['GpuHashAggregate'], 2))
+
+# Set minimum to 1.0 for speedup factors
+for key in scores_dict:
+    if float(scores_dict[key]) < min_speedup:
+        scores_dict[key] = f"{min_speedup}"
```

Comment:
Curious why we are setting scores_dict[key] to 1.0 instead of the number which we got earlier? 
If the cpu_stage_totals is less than gpu_stage_totals then does the above code imply that we are not accelerating that particular Exec but it is not slower than CPU run as well. So keeping it as 1.0?

Reply:
Yes, we're setting a minimum of 1.0 for an exec speedup as a safeguard.  If there is an environment where exec(s) are much slower on GPU, that would likely be a sign of a setup/configuration issue and I don't want that polluting qual tool results.  

I'm open to other ideas if you think we don't need the safeguard.